File: gems/decomplex/lib/decomplex/co_update.rb
# frozen_string_literal: true

require_relative "syntax"

module Decomplex
  # Co-update / inconsistent-update mining (cf. Lu et al., DynaMine).
  #
  # Plague targeted: "redundant state that drifts" and "the same path
  # in many places, one place misses the step." If attribute `.storage`
  # and attribute `.provenance` are assigned together in N methods, a
  # method that assigns one without the other is a probable desync --
  # exactly the documented invariant-#16 pairing whose violation is a
  # latent UAF.
  #
  # Unit of mutation = the ATTRIBUTE / ivar NAME (not the full receiver
  # chain): `node.storage =` and `decl.storage =` are the same logical
  # state edit, so they must cluster regardless of receiver. lvar
  # assignment is deliberately NOT mined (loop temps etc. are noise).
  #
  # Output is a ranked CANDIDATE list, not a verdict (Engler's
  # discipline): FP is acceptable, the receiver is printed so triage is
  # a one-line read.
  class CoUpdate
    Write = Struct.new(:attr, :recv, :file, :defn, :line, :span,
                       keyword_init: true)

    def self.scan(files)
      writes = files.flat_map do |file|
        Syntax.parse(file).state_writes.map do |write|
          Write.new(
            attr: write.field,
            recv: write.receiver,
            file: write.file,
            defn: write.function,
            line: write.line,
            span: write.span
          )
        end
      end
      Report.new(writes)
    end

    # Frequent co-written attribute pairs + the methods that break them.
    class Report
      def initialize(writes)
        @writes = writes
        @by_unit = writes.group_by { |w| [w.file, w.defn] }
      end

      # [{ pair:, support:, sites:[...] }, ...]
      def co_written_pairs(min_support: 3)
        counts = Hash.new { |h, k| h[k] = [] }
        @by_unit.each do |unit, ws|
          attrs = ws.map(&:attr).uniq.sort
          attrs.combination(2).each { |pair| counts[pair] << unit }
        end
        counts.filter_map do |pair, units|
          next if units.size < min_support

          { pair: pair, support: units.size,
            sites: units.map { |f, d| "#{f}:#{d}" } }
        end.sort_by { |h| -h[:support] }
      end

      # A method writing exactly one of a high-support pair. The other
      # attribute's absence next to a same-receiver write is the desync.
      # [{ pair:, support:, has:, missing:, at:, recv: }, ...]
      def neglected_updates(min_support: 3)
        pairs = co_written_pairs(min_support: min_support)
        out = []
        @by_unit.each do |(file, defn), ws|
          attrs = ws.map(&:attr).uniq
          pairs.each do |h|
            a, b = h[:pair]
            has, miss = if attrs.include?(a) && !attrs.include?(b)
                          [a, b]
                        elsif attrs.include?(b) && !attrs.include?(a)
                          [b, a]
                        end
            next unless has

            w = ws.find { |x| x.attr == has }
            out << { pair: h[:pair], support: h[:support], has: has,
                     missing: miss, at: "#{file}:#{defn}:#{w.line}",
                     spans: { "#{file}:#{defn}:#{w.line}" => w.span },
                     recv: w.recv }
          end
        end
        out.sort_by { |h| -h[:support] }
      end
    end
  end
end

File: gems/decomplex/rust/src/decomplex/detectors/co_update.rs
use crate::decomplex::ast::Span;
use crate::decomplex::syntax::{self, Language, StateWrite};
use anyhow::Result;
use serde::Serialize;
use std::collections::{BTreeMap, BTreeSet};
use std::path::PathBuf;

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct CoUpdateReport {
    pub co_written_pairs: Vec<CoWrittenPair>,
    pub neglected_updates: Vec<NeglectedUpdate>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct CoWrittenPair {
    pub pair: Vec<String>,
    pub support: usize,
    pub sites: Vec<String>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct NeglectedUpdate {
    pub pair: Vec<String>,
    pub support: usize,
    pub has: String,
    pub missing: String,
    pub at: String,
    pub spans: BTreeMap<String, Span>,
    pub recv: String,
}

#[derive(Clone, Debug)]
struct Write {
    attr: String,
    recv: String,
    file: String,
    defn: String,
    line: usize,
    span: Span,
}

pub fn scan_files(files: &[PathBuf], language: Language) -> Result<CoUpdateReport> {
    let mut writes = Vec::new();
    for file in files {
        let doc = syntax::parse_file(file.clone(), language)?;
        for w in doc.state_writes {
            writes.push(Write {
                attr: w.field,
                recv: w.receiver,
                file: w.file,
                defn: w.function,
                line: w.line,
                span: w.span,
            });
        }
    }
    let report = Report::new(writes);
    Ok(CoUpdateReport {
        co_written_pairs: report.co_written_pairs(3),
        neglected_updates: report.neglected_updates(3),
    })
}

pub fn state_writes_for_files(files: &[PathBuf], language: Language) -> Result<Vec<StateWrite>> {
    let mut out = Vec::new();
    for file in files {
        let doc = syntax::parse_file(file.clone(), language)?;
        out.extend(doc.state_writes);
    }
    Ok(out)
}

struct Report {
    #[allow(dead_code)]
    writes: Vec<Write>,
    by_unit: Vec<((String, String), Vec<Write>)>,
}

impl Report {
    fn new(writes: Vec<Write>) -> Self {
        let mut keys = Vec::new();
        let mut map: BTreeMap<(String, String), Vec<Write>> = BTreeMap::new();
        for w in &writes {
            let key = (w.file.clone(), w.defn.clone());
            if !map.contains_key(&key) {
                keys.push(key.clone());
            }
            map.entry(key).or_default().push(w.clone());
        }
        let by_unit = keys
            .into_iter()
            .map(|k| {
                let v = map.remove(&k).unwrap();
                (k, v)
            })
            .collect();
        Self { writes, by_unit }
    }

    fn co_written_pairs(&self, min_support: usize) -> Vec<CoWrittenPair> {
        let mut keys = Vec::new();
        let mut counts: BTreeMap<Vec<String>, Vec<(String, String)>> = BTreeMap::new();
        for (unit, ws) in &self.by_unit {
            let mut attrs: Vec<_> = ws
                .iter()
                .map(|w| w.attr.clone())
                .collect::<BTreeSet<_>>()
                .into_iter()
                .collect();
            attrs.sort();

            for i in 0..attrs.len() {
                for j in i + 1..attrs.len() {
                    let pair = vec![attrs[i].clone(), attrs[j].clone()];
                    if !counts.contains_key(&pair) {
                        keys.push(pair.clone());
                    }
                    counts.entry(pair).or_default().push(unit.clone());
                }
            }
        }

        let mut out = Vec::new();
        for pair in keys {
            let units = counts.remove(&pair).unwrap();
            if units.len() < min_support {
                continue;
            }
            out.push(CoWrittenPair {
                pair,
                support: units.len(),
                sites: units
                    .into_iter()
                    .map(|(f, d)| format!("{}:{}", f, d))
                    .collect(),
            });
        }
        out.sort_by(|a, b| b.support.cmp(&a.support));
        out
    }

    fn neglected_updates(&self, min_support: usize) -> Vec<NeglectedUpdate> {
        let pairs = self.co_written_pairs(min_support);
        let mut out = Vec::new();

        for ((file, defn), ws) in &self.by_unit {
            let attrs: BTreeSet<_> = ws.iter().map(|w| w.attr.clone()).collect();
            for p in &pairs {
                let a = &p.pair[0];
                let b = &p.pair[1];

                let (has, miss) = if attrs.contains(a) && !attrs.contains(b) {
                    (Some(a), Some(b))
                } else if attrs.contains(b) && !attrs.contains(a) {
                    (Some(b), Some(a))
                } else {
                    (None, None)
                };

                if let (Some(has), Some(miss)) = (has, miss) {
                    if let Some(w) = ws.iter().find(|x| &x.attr == has) {
                        let loc = format!("{}:{}:{}", file, defn, w.line);
                        let mut spans = BTreeMap::new();
                        spans.insert(loc.clone(), w.span);
                        out.push(NeglectedUpdate {
                            pair: p.pair.clone(),
                            support: p.support,
                            has: has.clone(),
                            missing: miss.clone(),
                            at: loc,
                            spans,
                            recv: w.recv.clone(),
                        });
                    }
                }
            }
        }
        out.sort_by(|a, b| b.support.cmp(&a.support));
        out
    }
}
